vopono_core/src/network/piapf.rs

```diff
@@ -0,0 +1,193 @@
+extern crate json;
+
+use std::sync::mpsc::{self, Receiver};
+use std::{
+    sync::mpsc::Sender,
+    thread::JoinHandle
+};
+use base64::prelude::*;
+use regex::Regex;
+
+use super::netns::NetworkNamespace;
+use super::Forwarder;
+
+use crate::config::vpn::Protocol;
+use crate::config::providers::OpenVpnProvider; // Added load_openvpn_auth to this trait 
+use crate::config::providers::pia::PrivateInternetAccess;  // Added load_wireguard_auth to this struct 
+
+/// Used to provide port forwarding for ProtonVPN
+pub struct Piapf {
+    pub port: u16,
+    loop_thread_handle: Option<JoinHandle<()>>,
+    send_channel: Sender<bool>,
+}
+
+struct ThreadParams {
+    pub port: u16,
+    pub netns_name: String,
+    pub signature: String, 
+    pub payload: String,
+    pub hostname: String,
+    pub gateway: String,
+}
+
+impl Piapf {
+    pub fn new(ns: &NetworkNamespace, protocol: &Protocol) -> anyhow::Result<Self> {
+        let pia = PrivateInternetAccess {}; //This is a bit weird, no? There's no state, so effectively all the methods are static...
+        
+        let traceroute_response = NetworkNamespace::exec_with_output(
+            &ns.name, 
+            &["traceroute", "-n", "-m", "1", "privateinternetaccess.com" ], )?;
+        if !traceroute_response.status.success() {
+            log::error!("Could not locate gateway with traceroute");
+            anyhow::bail!("Could not locate gateway with traceroute")
+        }
+        let re = Regex::new(r" *1 *(?P<gateway>\d{1,3}.\d{1,3}.\d{1,3}.\d{1,3}).*").unwrap();
+        let result = String::from_utf8_lossy(&traceroute_response.stdout);
+        let second_line = result.lines().skip(1).next().unwrap();
+        let vpn_gateway = re.captures(&second_line).unwrap().get(1).unwrap().as_str().to_string();
+        
+        log::info!("PIA gateway: {}", vpn_gateway);
+        
+        let vpn_hostname = match protocol {
+            Protocol::OpenVpn => "nl-amsterdam.privacy.network".to_string(), // FIXME: Parse this from the OpenVPN conf?
+            Protocol::Wireguard => "nl-amsterdam.privacy.network".to_string(), // FIXME: [Insert clever idea to get wireguard endpoint hostname here] 
```

Comment:
I think we can parse https://serverlist.piaservers.net/vpninfo/servers/v4 for both?

Reply:
For wireguard it was already parsing a v6 list which had the necessary info. Added it to the Config struct there.

Openvpn seems to come from PIA as a zip, and the files are named differently than the IDs in the v4/v6 lists, so created a similar map for these filenames by parsing out the endpoint. This will only work when the config files actually contain hostnames... the IP-only variants for now just won't support port forwarding. Perhaps one could parse the IPs then look them up in v4/v6...

Reply:
Thanks, I think it's okay for now!